vllm/model_executor/models/mixtral.py

```diff
@@ -66,8 +55,134 @@
+
+
+class MixtralMoE(nn.Module):
+
+    def __init__(
+        self,
+        config: MixtralConfig,
+        linear_method: Optional[LinearMethodBase] = None,
+    ):
+        super().__init__()
+        self.config = config
+        self.rank = get_tensor_model_parallel_rank()
+        self.tp_size = get_tensor_model_parallel_world_size()
+        self.num_total_experts = config.num_local_experts
+        self.top_k = config.num_experts_per_tok
+        if self.tp_size > self.num_total_experts:
+            raise ValueError(
+                f"Tensor parallel size {self.tp_size} is greater than "
+                f"the number of experts {self.num_total_experts}.")
+        # Split experts equally between ranks
+        self.expert_indicies = np.array_split(range(
+            self.num_total_experts), self.tp_size)[self.rank].tolist()
+        if not self.expert_indicies:
+            raise ValueError(
+                f"Rank {self.rank} has no experts assigned to it.")
+
+        self.experts = nn.ModuleList([
+            MixtralMLP(self.num_total_experts,
+                       config.hidden_size,
+                       config.intermediate_size,
+                       linear_method=linear_method)
+            if idx in self.expert_indicies else DummyModule()
+            for idx in range(self.num_total_experts)
+        ])
+        self.gate = ReplicatedLinear(config.hidden_size,
+                                     self.num_total_experts,
+                                     bias=False,
+                                     linear_method=linear_method)
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        batch_size, sequence_length, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.view(-1, hidden_dim)
+        # router_logits: (batch * sequence_length, n_experts)
+        router_logits, _ = self.gate(hidden_states)
+
+        routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+        routing_weights, selected_experts = torch.topk(routing_weights,
+                                                       self.top_k,
+                                                       dim=-1)
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+
+        final_hidden_states = None
+        for expert_idx in self.expert_indicies:
+            expert_layer = self.experts[expert_idx]
+            expert_mask = (selected_experts == expert_idx)
+            expert_weights = (routing_weights * expert_mask).sum(dim=-1,
+                                                                 keepdim=True)
+
+            current_hidden_states = expert_layer(hidden_states).mul_(
+                expert_weights)
+            if final_hidden_states is None:
+                final_hidden_states = current_hidden_states
+            else:
+                final_hidden_states.add_(current_hidden_states)
+
+        return tensor_model_parallel_all_reduce(final_hidden_states).view(
```

Comment:
Hi! I'm curious to understand what's going on in this implementation. The PR calls this expert parallelism but it still looks like tensor parallelism to me? At least, if this is expert parallelism, I don't see any logic routing the tokens to the device that owns the expert it was assigned to?

Reply:
We run every expert on the rank and mask out the rows that were not selected to be used by the expert. We then all reduce the tensors across the ranks. This results in dense computations (and higher memory usage), but it dramatically reduces latency, especially for small batch sizes.

Reply:
Ah ok, thank you! I was confused by the "expert parallelism" in the PR name, which I think is a misnomer here :)

The prior implementation with MegaBlocks was using a training-optimized code path. I'd expect it to be very inefficient because a) it pads each expert batch to the nearest multiple of 128 and b) dispatches to sparse matmul kernels which use tile dimensions tuned for large problems.

For inference, its much better to use our grouped implementation, which avoids these pitfalls. Essentially what is in the function [here](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/dmoe.py#L233). Our gather/scatter kernels handle replication for `top_k>1` as well as the permutation to group tokens by expert assignment. They're also written in Triton so they should work fine on AMD. For the MLP, we dispatch to custom grouped GEMM ops, but you can also use a pure-Torch grouped MLP like what's happening in this [PR](https://github.com/stanford-futuredata/megablocks/pull/42) to make it AMD compatible.

This is the direction I'd go to improve the current implementation further, fwiw. You don't necessarily need to add MegaBlocks as a dep - most of this can be replicated without too much complexity.

Reply:
Thanks for the explanation! I definitely agree there is a lot of room to expand here. Looking forward to more contributions from you or the community!